packages/adapter-components/src/client/pagination.ts

```diff
@@ -158,14 +159,19 @@ export const traverseRequests: (
  */
 export const getWithItemOffsetPagination = ({
   firstIndex,
-  itemsPerPage,
+  queryParamsPageSizeName,
 } : {
   firstIndex: number
-  itemsPerPage: number
+  queryParamsPageSizeName: string | undefined
 }): PaginationFunc => {
   const nextPage: PaginationFunc = ({ page, getParams, currentParams, pageSize }) => {
-    const { paginationField } = getParams
-    if (paginationField === undefined || page.length < pageSize) {
+    const { paginationField, queryParams } = getParams
+
+    const itemsPerPage = queryParamsPageSizeName !== undefined
+      && queryParams?.[queryParamsPageSizeName] !== undefined
+      ? Number(queryParams[queryParamsPageSizeName])
+      : pageSize
```

Comment:
1. readability nit: it starts becoming hard for readers to remember the order of operations, so it usually helps adding some parentheses, e.g.
```suggestion
    const itemsPerPage = (queryParamsPageSizeName !== undefined
      && queryParams?.[queryParamsPageSizeName] !== undefined)
      ? Number(queryParams[queryParamsPageSizeName])
      : pageSize
```
2. there are cases where `Number()` would result in `NaN` - let's add a fallback + tests for it (e.g. instead of checking `queryParams?.[queryParamsPageSizeName] !== undefined`, check `_.isNumber(Number(queryParams?.[queryParamsPageSizeName]))`